### ISSUE TITLE
ZFS: device or resource busy / dataset is busy

### ISSUE THREAD

**Issue by dsau** [docker/docker#24403]
**Output of `docker version`:**

```
Client:
 Version:      1.12.0-rc3
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   91e29e8
 Built:
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.0-rc3
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   91e29e8
 Built:
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 2
 Running: 0
 Paused: 0
 Stopped: 2
Images: 1
Server Version: 1.12.0-rc3
Storage Driver: zfs
 Zpool: tank
 Zpool Health: ONLINE
 Parent Dataset: tank/docker
 Space Used By Parent: 98304
 Space Available: 7621296459776
 Parent Quota: no
 Compression: lz4
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null host bridge overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 3.10.0-327.22.2.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 24
Total Memory: 125.7 GiB
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

The machine is physical, running Centos 7.2.1511 with ZoL v0.6.5.7-1.

**Steps to reproduce the issue:**
1. Start docker 1.12.0-rc3, with a ZFS volume mounted at /var/lib/docker
2. Create a simple Dockerfile
```
mkdir /tmp/docker-simple
cat << EOF > /tmp/docker-simple/Dockerfile
FROM alpine:latest
RUN echo testing > /tmp/testing
EOF
```
3. Build it
```
docker build -t local/simple /tmp/docker-simple
```

**Describe the results you received:**
```
Sending build context to Docker daemon 2.048 kB
Step 1 : FROM alpine:latest
 ---> 4e38e38c8ce0
Step 2 : RUN echo testing > /tmp/testing
 ---> Running in 7553774d9301
error creating zfs mount of tank/docker/73360b9c302e91614ac2b1727d161fb8fdc42b49903f8211ed4356f0fecf8b8b to /var/lib/docker/zfs/graph/73360b9c302e91614ac2b1727d161fb8fdc42b49903f8211ed4356f0fecf8b8b: device or resource busy
```

**Describe the results you expected:**
When performing the same actions on 1.12.0-0.2.rc2 the build completes successfully.
```
Sending build context to Docker daemon 2.048 kB
Step 1 : FROM alpine:latest
 ---> 4e38e38c8ce0
Step 2 : RUN echo testing > /tmp/testing
 ---> Running in a518cd368625
 ---> 09a63abb1159
Removing intermediate container a518cd368625
Successfully built 09a63abb1159
```

**Additional information you deem important (e.g. issue happens only occasionally):**
The issue occurs on every attempt, I've also noticed that I cannot start the docker daemon on RC3 until I manually create the /var/lib/docker/zfs directory.
```
DEBU[0001] [zfs] ID:cab1bc32-09bb-47d1-b73b-80f48a694dc5 START zfs list -rHp -t filesystem -o name,origin,used,available,mountpoint,compression,type,volsize,quota,written,logicalused,usedbydataset tank/docker
DEBU[0001] [zfs] ID:cab1bc32-09bb-47d1-b73b-80f48a694dc5 FINISH
DEBU[0001] Cleaning up old mountid : start.
FATA[0001] Error starting daemon: error initializing graphdriver: no such file or directory
```
It seems to mount the /var/lib/docker/zfs directory to the same zfs volume as /var/lib/docker, a process that wasn't occurring in previously. Output from RC2,
```
# mount | grep docker
tank/docker on /var/lib/docker type zfs (rw,noatime,seclabel,xattr,noacl)
```
and in RC3
```
# mount | grep docker
tank/docker on /var/lib/docker type zfs (rw,noatime,seclabel,xattr,noacl)
tank/docker on /var/lib/docker/zfs type zfs (rw,noatime,seclabel,xattr,noacl)
```

A similar issue occurs after a container is removed, however this is intermittent.
```
# docker run -it --rm alpine ping -c 1 8.8.8.8
PING 8.8.8.8 (8.8.8.8): 56 data bytes
64 bytes from 8.8.8.8: seq=0 ttl=60 time=1.234 ms

--- 8.8.8.8 ping statistics ---
1 packets transmitted, 1 packets received, 0% packet loss
round-trip min/avg/max = 1.234/1.234/1.234 ms
Error response from daemon: Driver zfs failed to remove root filesystem 208e64b201ea674d56d6028882f45a574aaa31dc2826ea21eaaa204b85161564: exit status 1: "/sbin/zfs zfs destroy -r tank/docker/ecd5729475a065a7d3ef71e1cdfef73c2e09e4099faf37ab12e1bc5274b78bc7" => cannot destroy 'tank/docker/ecd5729475a065a7d3ef71e1cdfef73c2e09e4099faf37ab12e1bc5274b78bc7': dataset is busy
```
```
DEBU[0051] Calling DELETE /v1.24/containers/208e64b201ea674d56d6028882f45a574aaa31dc2826ea21eaaa204b85161564?force=1&v=1
DEBU[0051] [zfs] ID:24f591d3-6ec2-4bdd-b502-be6e55ea1e6d START zfs destroy -r tank/docker/ecd5729475a065a7d3ef71e1cdfef73c2e09e4099faf37ab12e1bc5274b78bc7
DEBU[0051] [zfs] ID:24f591d3-6ec2-4bdd-b502-be6e55ea1e6d FINISH
ERRO[0051] Error removing mounted layer 208e64b201ea674d56d6028882f45a574aaa31dc2826ea21eaaa204b85161564: exit status 1: "/sbin/zfs zfs destroy -r tank/docker/ecd5729475a065a7d3ef71e1cdfef73c2e09e4099faf37ab12e1bc5274b78bc7" => cannot destroy 'tank/docker/ecd5729475a065a7d3ef71e1cdfef73c2e09e4099faf37ab12e1bc5274b78bc7': dataset is busy

ERRO[0051] Handler for DELETE /v1.24/containers/208e64b201ea674d56d6028882f45a574aaa31dc2826ea21eaaa204b85161564 returned error: Driver zfs failed to remove root filesystem 208e64b201ea674d56d6028882f45a574aaa31dc2826ea21eaaa204b85161564: exit status 1:
"/sbin/zfs zfs destroy -r tank/docker/ecd5729475a065a7d3ef71e1cdfef73c2e09e4099faf37ab12e1bc5274b78bc7" => cannot destroy 'tank/docker/ecd5729475a065a7d3ef71e1cdfef73c2e09e4099faf37ab12e1bc5274b78bc7': dataset is busy
```

**Comment by cpuguy83** [comment]
I cannot reproduce this.
Just setup a fresh install (though ubuntu since that's what I had available), setup zfs and it all just worked, didn't have to create any dirs or anything.

Do you have any more details on how I can reproduce?

**Comment by cpuguy83** [comment]
How did you setup zfs? Docker?

**Comment by dsau** [comment]
Hmmm, I'm running ZFS as my root partition, I wonder if that has something to do with it.

I've just set up a centos VM and gone through the following steps.

```
su -
yum upgrade -y
yum localinstall -y --nogpgcheck http://archive.zfsonlinux.org/epel/zfs-release.el7.noarch.rpm
sed -i 's|7\/$basearch|7\/kmod\/$basearch|' /etc/yum.repos.d/zfs.repo
yum install -y zfs
modprobe zfs
dd if=/dev/zero of=/tmp/tank.img bs=1k count=1 seek=100M
zpool create tank /tmp/tank.img
curl -fsSL https://test.docker.com/ | sh
systemctl start docker
```
So everything looks fine, even without creating the directories.
```
[root@docker-test ~]# docker run -it --rm alpine echo it works!
Unable to find image 'alpine:latest' locally
...
Status: Downloaded newer image for alpine:latest
it works!

[root@docker-test ~]# docker info
...
Server Version: 1.12.0-rc3
Storage Driver: devicemapper
 Pool Name: docker-253:1-117499930-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
...
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Use `--storage-opt dm.thinpooldev` to specify a custom block storage device.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.107-RHEL7 (2016-06-09)
...
[root@docker-test ~]# zfs list
NAME   USED  AVAIL  REFER  MOUNTPOINT
tank   103K  96.4G    19K  /tank
```
Ah, it works, however it's not using the ZFS pool.
```
systemctl stop docker
yum remove -y docker-*
rm -rf /var/lib/docker
zfs create -o mountpoint=/var/lib/docker tank/docker
yum install -y docker-engine

[root@docker-test ~]# systemctl start docker
Job for docker.service failed because the control process exited with error code. See "systemctl status docker.service" and "journalctl -xe" for details.
```
Same error, I am receiving on the physical host, doesn't want to start until I give it the zfs folder...
```
[root@docker-test ~]# mkdir /var/lib/docker/zfs
[root@docker-test ~]# systemctl start docker
[root@docker-test ~]# docker run -it --rm alpine echo it works!
Unable to find image 'alpine:latest' locally
...
Status: Downloaded newer image for alpine:latest
it works!

[root@docker-test ~]# docker info
...
Server Version: 1.12.0-rc3
Storage Driver: zfs
 Zpool: tank
 Zpool Health: ONLINE
 Parent Dataset: tank/docker
...
 Parent Quota: no
 Compression: off
Logging Driver: json-file
Cgroup Driver: cgroupfs
...
[root@docker-test ~]# zfs list
NAME                                                                           USED  AVAIL  REFER  MOUNTPOINT
tank                                                                          8.45M  96.4G    19K  /tank
tank/docker                                                                   7.93M  96.4G  2.56M  /var/lib/docker
tank/docker/78299ad01b1cdb9ce56aa3a83d11f6715aa44e8d6e2b966b8ce3e028fd68a98e  5.38M  96.4G  5.38M  legacy
```
Ok, so that works and is using the ZFS pool.  I tried my build script from above, and that also works.

Went back and tried it all again on the physical host, and nope, no good. There is definitely something different in relation to the graphdriver with RC3, as I never had to create the zfs folder previously. Also, the physical host functions perfectly on RC2.

I'll look into running ZFS for the root partition within a VM and see if that sheds any light.

**Comment by dsau** [comment]
Ok, I think I've tracked it down. It's related to the mount at /var/lib/docker/zfs which appears when starting the daemon.
```
[root@physical ~]# zfs create -o mountpoint=/var/lib/docker tank/docker
[root@physical ~]# yum install -y docker-engine
[root@physical ~]# mkdir /var/lib/docker/zfs
[root@physical ~]# systemctl start docker

[root@physical ~]# mount | grep docker
tank/docker on /var/lib/docker type zfs (rw,noatime,seclabel,xattr,noacl)
tank/docker on /var/lib/docker/zfs type zfs (rw,noatime,seclabel,xattr,noacl)
[root@physical ~]# docker run -it --rm alpine echo it works!
Unable to find image 'alpine:latest' locally
latest: Pulling from library/alpine
e110a4a17941: Pull complete
Digest: sha256:3dcdb92d7432d56604d4545cbd324b14e647b313626d99b889d0626de158f73a
Status: Downloaded newer image for alpine:latest
it works!
Error response from daemon: Driver zfs failed to remove root filesystem 277aa33496a435277fb0f7a260ccf7466c0e42c82c93d2160c43424dbe4c872a: exit status 1: "/usr/sbin/zfs zfs destroy -r tank/docker/57d356fcdcb191c731fa8b2ccd8f738a55128d817bd65dda97a10f3b050a98e5" => cannot destroy 'tank/docker/57d356fcdcb191c731fa8b2ccd8f738a55128d817bd65dda97a10f3b050a98e5': dataset is busy
[root@physical ~]#
```
With that mount in place, I constantly receive errors whenever a mount or destroy is attempted on a zfs volume.

However, when removing it after launch but before performing any tasks...
```
[root@physical ~]# systemctl stop docker
[root@physical ~]# umount /var/lib/docker/zfs
[root@physical ~]# zfs destroy -R tank/docker
[root@physical ~]# rm -rf /var/lib/docker
[root@physical ~]# zfs create -o mountpoint=/var/lib/docker tank/docker
[root@physical ~]# yum reinstall docker-engine
[root@physical ~]# mkdir /var/lib/docker/zfs
[root@physical ~]# systemctl start docker
[root@physical ~]# mount | grep docker
tank/docker on /var/lib/docker type zfs (rw,noatime,seclabel,xattr,noacl)
tank/docker on /var/lib/docker/zfs type zfs (rw,noatime,seclabel,xattr,noacl)
[root@physical ~]# umount /var/lib/docker/zfs
[root@physical ~]# mount | grep docker
tank/docker on /var/lib/docker type zfs (rw,noatime,seclabel,xattr,noacl)
[root@physical ~]# docker run -it --rm alpine echo it works!
Unable to find image 'alpine:latest' locally
latest: Pulling from library/alpine
e110a4a17941: Pull complete
Digest: sha256:3dcdb92d7432d56604d4545cbd324b14e647b313626d99b889d0626de158f73a
Status: Downloaded newer image for alpine:latest
it works!
[root@physical ~]#
```
Everything works perfectly. I've double checked and that additional mount only started appearing in 1.12.RC3, wasn't there in RC2.

**Comment by cpuguy83** [comment]
That's interesting, this is likely coming from 5ec8441bb55ecbf83da6407a4135ce83b1fe01c7

**Comment by cpuguy83** [comment]
I can't reproduce the original issue that the above commit fixed (w/o the change, of course).
ping @AkihiroSuda 

@dsau I'm also wondering if on one system you are using fuse and the other you are using ZoL? Just trying to narrow down the differences.

**Comment by AkihiroSuda** [comment]
@cpuguy83 
You mean,
- you did not hit https://github.com/docker/docker/issues/24008 in RC2
- you did not hit this issue #24403 in RC3,

right?

When I hit #24008 in RC2, I was using Ubuntu 16.04, and did `apt install zfs` and `zfs create -o mountpoint=/var/lib/docker zpool-docker/docker` for setting up zfs. (https://github.com/docker/docker/blob/d0737e9ac0bfcbac0e212d157ab305e561eea3ee/docs/userguide/storagedriver/zfs-driver.md)
I did not use systemd at that time.
If you were using systemd and had `Mountflags` in your local `/lib/init/systemd/docker.service`, probably it affects the reproducibility of the issue. (2aee081cad72352f8b0c37ba0414ebc925b022e8)

I'm trying to hit this issue #24403, but did not hit yet (in the master commit 8f3be176d26948dfac9b6d855034dabae78454be, Jul 11)

**Comment by cpuguy83** [comment]
@AkihiroSuda I mean I can not hit #24008 when I take out your patch to make the zfs base private.
I set my system up the same way (and using 16.04 with the zfs package).
I *am* using systemd, but the unit file is no longer using `MountFlags`.

**Comment by AkihiroSuda** [comment]
I confirmed I can hit #24008 when I excluded https://github.com/docker/docker/commit/5ec8441bb55ecbf83da6407a4135ce83b1fe01c7, and can avoid  it when I included the commit.
@cpuguy83 Are you sure you added `--no-cache` when you tried to test #24008?

**Comment by cpuguy83** [comment]
@AkihiroSuda Yep. I'm not saying the patch is wrong, but just trying to repro any issue and not having any luck :(

**Comment by dsau** [comment]
@cpuguy83 I'm using ZoL on all machines, v0.6.5.7-1.

I experience the same issue on two identical bare-metal servers, but not in my local VM environment, or, on two other slightly different bare-metal hosts. It's a weird one.

The main difference that comes to mind is the disks, 7.2k SATA and the problem surfaces, all the other hosts, both physical and virtual are backed by either SSD/15k SAS, not sure if this is at all relevant, perhaps the increased latency could be causing a deadlock? Wild stab in the dark.

I'll try rebuild one of the hosts which are experiencing the issue (was hoping it didn't come to that). But I'm not holding my breath as they were both recently built and are fairly clean.

Is the mount at /var/lib/docker/zfs expected?  At least in my setup, where I already have /var/lib/docker mounted to a zfs dataset, it doesn't appear to provide any purpose. I'm happy to just drop "ExecStartPost=umount /var/lib/docker/zfs" in my docker.service file and call it a day if it's not impacting anyone else.

**Comment by cpuguy83** [comment]
@dsau This is a re-mount that makes the mountpoint private so mounts don't leak into other namespaces.

**Comment by cpuguy83** [comment]
I'm curious why we are using legacy mountpoints anyway.

**Comment by dsau** [comment]
@cpuguy83 I've rebuilt one of the physical hosts that were having the issue, and I can no longer reproduce it. I'll spend some time on it in the morning and try to figure out what is the difference between these two servers.

**Comment by cpuguy83** [comment]
@dsau Thanks for looking deeper, I have not been able to reproduce.

**Comment by dsau** [comment]
@cpuguy83

I've been able to reproduce this on all servers. Starting from a fresh, minimal install of CentOS 7, here is the least amount of config I could find which would cause the failure.
```
sudo su -

yum -y localinstall --nogpgcheck http://archive.zfsonlinux.org/epel/zfs-release.el7.noarch.rpm
sed -i 's/7\/$basearch/7\/kmod\/$basearch/' /etc/yum.repos.d/zfs.repo
yum -y install zfs systemd-networkd systemd-resolved && modprobe zfs

mkdir /etc/systemd/network; cat << EOF > /etc/systemd/network/eth0.network
[Match]
Name=eth0
[Network]
Address=$(ip addr show eth0 | grep "inet " | awk '{print $2}')
Gateway=$(ip route | grep "default" | awk '{print $3}')
DNS=8.8.8.8
EOF

dd if=/dev/zero of=/tmp/tank.img bs=1k count=1 seek=100M && zpool create tank /tmp/tank.img
zfs create -o mountpoint=/var/lib/docker tank/docker && mkdir /var/lib/docker/zfs

curl -fsSL https://test.docker.com/ | sh

systemctl mask NetworkManager network
systemctl enable systemd-networkd systemd-resolved docker
ln -sf /run/systemd/resolve/resolv.conf /etc/resolv.conf

mkdir /tmp/docker-simple; cat << EOF > /tmp/docker-simple/Dockerfile
FROM alpine:latest
RUN echo testing > /tmp/testing
EOF

shutdown -r now
```
The host has to be restarted, couldn't get it to fail without it.
```
[root@localhost ~]# docker build -t local/simple /tmp/docker-simple
Sending build context to Docker daemon 2.048 kB
Step 1 : FROM alpine:latest
latest: Pulling from library/alpine
e110a4a17941: Pull complete
Digest: sha256:3dcdb92d7432d56604d4545cbd324b14e647b313626d99b889d0626de158f73a
Status: Downloaded newer image for alpine:latest
 ---> 4e38e38c8ce0
Step 2 : RUN echo testing > /tmp/testing
 ---> Running in 7b8e68d05e7d
error creating zfs mount of tank/docker/3b2fcaadffd82e5b5098809dcf4840b20d50dab45bc34778729034f4ed183714 to /var/lib/docker/zfs/graph/3b2fcaadffd82e5b5098809dcf4840b20d50dab45bc34778729034f4ed183714: device or resource busy
```
If it doesn't happen on the first attempt, just keep clearing the data, restarting docker and trying again.

```
systemctl stop docker; umount /var/lib/docker/zfs; zfs destroy -R tank/docker
zfs create -o mountpoint=/var/lib/docker tank/docker && mkdir /var/lib/docker/zfs
systemctl start docker

docker build -t local/simple /tmp/docker-simple
```

**Comment by cpuguy83** [comment]
Still scratching my head on this one...

**Comment by dsau** [comment]
@cpuguy83  yep, it's a strange one, were you able to replicate?

I may have some info that can help. Seems to be a combination of zfs, networkd, and the private mount at /var/lib/docker/zfs.

As of systemd v.219, networkd shuts itself down if it's idle for 30 seconds. It's also got a systemd socket, to wake it up in case anything happens that it needs to handle.

So if I wait for networkd to sleep, then attempt something which will wake it up, and perform a zfs destroy in quick succession, i.e. a build or short-lived run with --rm, it fails every time. But, do the same while it's still active, and it succeeds.

Disabling systemd-networkd.socket causes networkd to sleep through any changes docker performs to the network, and the issue doesn't occur. However, that would also cause it to sleep through a carrier loss on one of the bonded nics, so I can't do that.

If I force the private mount at /var/lib/docker/zfs to be shared with 'MountFlags=shared' in the docker.service file, then it's fine, running the zfs destroy while networkd is waking up, is successful, no 'dataset is busy' errors.

So remove any of the pieces and it's fine. Seems like a race condition between networkd waking up, and the umount/zfs destroy.

**Comment by cpuguy83** [comment]
@dsau I cannot repro.
The only way I can see this being a problem is if mounts have already leaked out, but I'm not sure why making the docker zfs root private would cause a problem there.
I'm not all that familiar with ZFS, though.... could somehow the system dataset be the same dataset used for containers?

**Comment by AkihiroSuda** [comment]
I succeeded to reproduce this issue on GCE CentOS7 image.
I'm looking into this again.
The steps are almost identical to https://github.com/docker/docker/issues/24403#issuecomment-232641827, but I did _not_ have to execute following:
```
mkdir /etc/systemd/network; cat << EOF > /etc/systemd/network/eth0.network
[Match]
Name=eth0
[Network]
Address=$(ip addr show eth0 | grep "inet " | awk '{print $2}')
Gateway=$(ip route | grep "default" | awk '{print $3}')
DNS=8.8.8.8
EOF
```
```
systemctl mask NetworkManager network
systemctl enable systemd-networkd systemd-resolved docker
ln -sf /run/systemd/resolve/resolv.conf /etc/resolv.conf
```

**Comment by AkihiroSuda** [comment]
- `for f in $(seq 1 10);do docker build -q --no-cache -t local/simple /tmp/docker-simple; sleep 25; done`: 100% succeeds
- `for f in $(seq 1 10);do docker build -q --no-cache -t local/simple /tmp/docker-simple; sleep 35; done`: 100% fails

seems definitely related to the networkd's 30-second thing

**Comment by dsau** [comment]
Awesome, thanks for that @AkihiroSuda, I was beginning to think I was going crazy. It looks like the networkd idle detection was reworked in systemd/systemd@9dc907f.

I'm still a little confused as to why a socket triggering a launch of networkd would prevent ZFS from destroying a dataset. Maybe an issue in the mount handling while a service is being launched?

@cpuguy83, which versions of systemd-networkd have you tried? I can provide a clean Centos7 VM with the issue occurring if that would help?

However, it seems like this one may need fixing elsewhere.

**Comment by ikogan** [comment]
I'm seeing this issue, it seems, as well. I've posted some investigation (including system information) in https://github.com/docker/docker/issues/16112#issuecomment-238932212. Mounts do appear to be leaking at some point, but I'm not quite sure when or how the processes are related.

I have httpd setup to proxy to the container being deleted over the internal docker bridge and named automatically attaches to the interface on the mariadb container.

**Comment by Mic92** [comment]
@cpuguy83 I used legacy mountflags, so zfs does not try to mount all zfs datasets at pool import time, which is slow on linux. (zfs mount requires the kernel to call userland mount due GPL-licenced  `vfs_mount` api). Also other programs daemon with mounts gets slow if a lot of docker layer are mounted at the same time.

**Comment by Mic92** [comment]
Leaking mountpoints might not be the only reason, why destroying fails, it could be also that one process still have open files on this filesystem.
A workaround might be this patch:

```
https://github.com/Mic92/docker/commit/31025581db4863434b5105a4acfc084200f34185
```

A similar approach was also proposed by zfsonlinux maintainer for [`zfs destroy`](https://github.com/zfsonlinux/zfs/issues/2435)

**Comment by cpuguy83** [comment]
We hit this consistently in one of our CI boxes (on zfs) so I'll be doing some more testing there.

**Comment by cfstras** [comment]
I think I'm also hitting this on my builds.
Builds are triggered through a gitlab-runner docker image (has docker.sock mounted), so that may be an additional variable.

docker v1.12.1-1, zfs-dkms 0.6.5.7-1, linux 4.4.19-1-lts on Arch Linux.

Log mentions multiple failures:
```
Error removing intermediate container 053ec5a23b1b: Driver zfs failed to remove root filesystem 053ec5a23b1b221cda8899c2e062401905cb590d05dd83218ded9d7f51a0a91e: exit status 1: "/usr/sbin/zfs zfs destroy -r tank/docker/91b22f1f2375b828fe80f171161e8088306760d29927ecb7b092a8533e0a06dc" => cannot destroy 'tank/docker/91b22f1f2375b828fe80f171161e8088306760d29927ecb7b092a8533e0a06dc': dataset is busy
Error removing intermediate container 053ec5a23b1b: No such container: 053ec5a23b1b221cda8899c2e062401905cb590d05dd83218ded9d7f51a0a91e
(repeated on other steps)

and finally 
error creating zfs mount of tank/docker/b022faf135592418802f166cc47bcab94f58867ddb6e4af876440deae2fcf8f7 to /var/lib/docker/zfs/graph/b022faf135592418802f166cc47bcab94f58867ddb6e4af876440deae2fcf8f7: device or resource busy
```
https://gist.github.com/cfstras/ef74ed97a001ef4f0fe85d6edc6e886c

It's not 100% reproducible, sometimes removing the base images and intermediate containers fixes it...

**Comment by Mic92** [comment]
I have opened a feature request for deferred deletes on zfsonlinux, which could solve the problem:

https://github.com/zfsonlinux/zfs/issues/5038

devicemapper already has such an option.

**Comment by nikarul** [comment]
Has there been any progress on this?  I ran into this issue on my local developer environment, my docker-compose setup fails and certain steps due to create/unmount failures.

**Comment by andrewbrs** [comment]
Hopefully relevant, I've also been experiencing something at least similar to this for some time.

Recently I discovered the mount looks to be held in an unrelated processes mount namespace, usually at least in my case ntpd (but have seen it in others)?... 

As a result restarting whichever process the mount seems to have leaked into, frees the filesystem and allow the action to work:

```
# docker rm acontainer
Error response from daemon: Driver zfs failed to remove root filesystem a3833feae8c4c7094de704905cd11b664685aef2c72e43374666344ef25e404c: exit status 1: "/sbin/zfs zfs destroy -r tank/system/data/v
ar/lib/docker/47a7f24ac0cd355f8d0ac2d022c38db7314b1b0f7afe8b24cbcecec37fa75684" => cannot destroy 'tank/system/data/var/lib/docker/47a7f24ac0cd355f8d0ac2d022c38db7314b1b0f7afe8b24cbcecec37fa75684':
 dataset is busy

# grep -Hir 47a7f24a /proc/*/mountinfo
/proc/8883/mountinfo:410 408 0:193 / /var/lib/docker/zfs/graph/47a7f24ac0cd355f8d0ac2d022c38db7314b1b0f7afe8b24cbcecec37fa75684 rw,relatime shared:217 - zfs tank/system/data/var/lib/docker/47a7f24ac0cd355f8d0ac2d022c38db7314b1b0f7afe8b24cbcecec37fa75684 rw,xattr,noacl

# ps -p 8883
  PID TTY          TIME CMD
 8883 ?        00:01:55 ntpd

# systemctl restart ntpd.service

# grep -Hir 47a7f24a /proc/*/mountinfo

# docker rm acontainer
acontainer
```

Still a bit unfamiliar with the finer details of linux mount namespaces, but did note my ntpd under systemd does some mount namespacing itself (```PrivateTmp=true```) but unsure if that is of any relevance to the issue:

```
# systemctl cat ntpd
# /usr/lib64/systemd/system/ntpd.service
[Unit]
Description=Network Time Service
After=ntpdate.service sntp.service
Conflicts=systemd-timesyncd.service

[Service]
ExecStart=/usr/sbin/ntpd -g -n
PrivateTmp=true

[Install]
WantedBy=multi-user.target

# /etc/systemd/system/ntpd.service.d/after-wan-online.conf
[Unit]
After=network-wan-online.target
Wants=network-wan-online.target
```

Hope this is of some help...

**Comment by Mic92** [comment]
@andrewbrs yes. this is a known issue, I also with experienced with LXC. Systemd creates a new namespace without mount namespace propragation, which inherits mounts from the host.
When zfs the commandline application unmounts the dataset in the parent namespace, this information is not propagated to its childs. In turn zfs cannot destroy the filesystem, because it is still mounted.
I think proper mount namespace support has to be implemented in zfs to get rid of those bugs. The only current workaround as a user would be to restart the services in questions. The only workaround for docker would be to retry destroying datasets.

**Comment by syb** [comment]
After a new install of Ubuntu 16.10 (with root on ZFS) and manual creation of ZFS filesystem at `/var/lib/docker`, then in initial testing I've had the same "dataset is busy" problem preventing container deletion.
- docker-engine: 1.12.5-0~ubuntu-xenial; 
- zfs-initramfs: 0.6.5.8-0ubuntu4.1
- My ZFS dataset structure may be similar to @dsau because I too have `/var/lib/docker/zfs` directory in the same zfs filesystem dataset as `/var/lib/docker`.

---
My understanding of the state if this issue is that Docker (and sometimes LXC) is exposing a design flaw or low-level bug :bug: in interaction between *SystemD* and *ZFS*, where "mounts are leaking between namespaces". Two questions to try progress work on this issue:-

1. Would it be helpful getting the right people working on the bug, if I try repro on previous Ubuntu that doesn't use SystemD? :question: 
1. Could Docker code be enhanced to debug log which namespaces a ZFS filesystem is mounted before attempting the 'zfs destory' command?

**Comment by cpuguy83** [comment]
@syb This goes beyond systemd or zfs.
Docker is creating mounts in it's namespace for containers.
containerD and runC (must, currently) share a mount namespace with dockerd.
runC also mounts stuff in this namespace.

Typically dockerd has the same mount namespace as the base system so a user who mounts stuff onto the host can then subsequently mount stuff into the container.

This all leads some really bad behavior.

If a user is not going to be mounting anything on the host or use v1 plugins then docker can be fired up with it's own private mount namespace (shared with runc and containerd only) with `MountFlags=private`, which would improve things significantly.
This effectively makes systemd give docker it's own mount namespace, and then sets `/` to private.

If a user does want to mount things then they can set use systemd to with `MountFlags=slave`. Like above, but sets `/` to slave so mounts propagate from the host namespace to docker's namespace, but not the other way around.
I believe if someone is using v2 plugins only (no v1) `slave` could be used pretty safely.
This should also improve things.

Ideally dockerd would setup it's own mount namespace (this is tricky in go) and also have a separate process to create mount namespaces for container that gets passed to runC... but runc does not currently support this.

**Comment by shapiroj** [comment]
Is it possible to set the MountFlags if I'm running dockerd directly and not using systemd?

**Comment by cpuguy83** [comment]
@shapiroj You can use `unshare -m dockerd`, but this isn't quite the same.
MountFlags is doing two things, creating a new mount namespace (this is what `unshare -m` does) and then setting the root mount `/` to whatever value you set in mount flags... e.g. `mount --make-private /`

Maybe something like `unshare -m sh -c 'mount --make-private / && exec dockerd'`

**Comment by Mic92** [comment]
@cpuguy83 should it not be `unshare -m sh -c 'mount --make-rprivate / && exec dockerd'` ?

**Comment by cpuguy83** [comment]
Maybe? This would probably only effect existing mounts, not new ones... but I'm definitely not positive what recursive really means in this context.

**Comment by Mic92** [comment]
@cpuguy83 recursive mean all mounts below "/", which is effectively every mount. New mounts on private mounts will become private too: https://www.kernel.org/doc/Documentation/filesystems/sharedsubtree.txt 

```
2c) A private mount does not forward or receive propagation.
```

**Comment by shapiroj** [comment]
I tried this and still get the 'dataset is busy' errors.   I actually had to use unshare -m -- mount --make-rprivate / && exec dockerd, as the syntax given above does not work.  I'm pretty sure dockerd is running in its own mount namespace, as I can no longer see from the host the mounts being created in /var/lib/docker/zfs for each container.

**Comment by Mic92** [comment]
So processes in the mount namespace still use the filesystem. It is not only `zfs`, which suffer from this problem. Maybe adding general support for deferred deletion in the docker daemon would solve this problem for every filesystem.

**Comment by viawest-davidsix** [comment]
I'm still seeing regular `error creating zfs mount...device or resource busy` errors during build even after setting `MountFlags=private`
Is there some other configuration workaround for this issue?

**Comment by Mic92** [comment]
This is a race condition as far as I know, some processes still has open file handles. I do not know which one.

**Comment by viawest-davidsix** [comment]
Would it be worthwhile to have an option to retry the operation a limited number of times instead of failing the build? I have never seen a repeat attempt to build the same image fail.

**Comment by cpuguy83** [comment]
@viawest-davidsix When you get the busy error and you manually try to unmount, does it unmount properly?

**Comment by viawest-davidsix** [comment]
I usually see the issue with automated builds, so I am not on the machine when the error occurs. I'm running some tests to see if I can find a reliable way to reproduce the issue so that I can test the mount operation manually.

That said, I have never seen the re-run of a failed automated job fail. (granted this may be a few minutes later)

To confirm, if the error I get is:
`error creating zfs mount of docker/lib/c4fb23bc7c4b4d4fb71ca2b402d2b2fb152912593a786dbe42171abcda7b3f69 to /var/lib/docker/zfs/graph/c4fb23bc7c4b4d4fb71ca2b402d2b2fb152912593a786dbe42171abcda7b3f69: device or resource busy`
Then the command to test would be:
`mount -t zfs docker/lib/c4fb23bc7c4b4d4fb71ca2b402d2b2fb152912593a786dbe42171abcda7b3f69 /var/lib/docker/zfs/graph/c4fb23bc7c4b4d4fb71ca2b402d2b2fb152912593a786dbe42171abcda7b3f69`
Correct?

**Comment by cpuguy83** [comment]
Seems right, assuming there's a zfs mount helper.

Seems strange though that it would cause EBUSY unless maybe the dataset is already mounted?

**Comment by dustinvanbuskirk** [comment]
We keep seeing this on container restarts.  

I've used fuser and found the docker process is still using the mount and sees the mount as busy.  The only workaround we've found is to create a new container which isn't much of a workaround.

**Comment by Mic92** [comment]
@dustinvanbuskirk can spot, which files are in use?

**Comment by dustinvanbuskirk** [comment]
@Mic92 Looks like I spoke too soon.  It looks like the behavior I am experiencing is not the same zfs issue before.  @cpuguy83 seems to be on to something... 

``` console
root@host:~# zfs get mounted zpool-docker/docker/b279ce094c0a6f455ad2d7c6ed63558de041e2426f8c768dcf44a2fb76a8e65a
NAME                                                                                  PROPERTY  VALUE    SOURCE
zpool-docker/docker/b279ce094c0a6f455ad2d7c6ed63558de041e2426f8c768dcf44a2fb76a8e65a  mounted   no       -
root@host:~# mount -t zfs zpool-docker/docker/b279ce094c0a6f455ad2d7c6ed63558de041e2426f8c768dcf44a2fb76a8e65a /var/lib/docker/zfs/graph/b279ce094c0a6f455ad2d7c6ed63558de041e2426f8c768dcf44a2fb76a8e65a
filesystem 'zpool-docker/docker/b279ce094c0a6f455ad2d7c6ed63558de041e2426f8c768dcf44a2fb76a8e65a' is already mounted
root@host:~# zfs get mounted zpool-docker/docker/b279ce094c0a6f455ad2d7c6ed63558de041e2426f8c768dcf44a2fb76a8e65a
NAME                                                                                  PROPERTY  VALUE    SOURCE
zpool-docker/docker/b279ce094c0a6f455ad2d7c6ed63558de041e2426f8c768dcf44a2fb76a8e65a  mounted   no       -```

**Comment by thaJeztah** [comment]
Yes, cAdvisor is known to cause device busy issues as it mounts `/var/lib/docker`

**Comment by devx** [comment]
I can validate @dustinvanbuskirk finding, and i would also agree that it's not a zfs issue.  As I ran into the same issue. Once i stopped ~cadvisor~ everything worked cleanly.

**Comment by jdrowell** [comment]
I just had this issue for the first time, and arrived here because I also use zfs (which seems to be unrelated to the issue at hand). But in my case, it was bluetoothd who was holding the mount (I ran it to test a gamepad). Yes, weird. So in my case, EBUSY is indeed accurate for a docker restart, and the real problem seems to be other long running processes that were started during the previous run of the container.

**Comment by rueberger** [comment]
Anecdata in support of a race condition problem: I occasionally experience this issue while the docker engine is trying to restart broken services (so just a restart loop). When I kill the broken service, I stop experiencing the issue.

**Comment by chrisstornowski** [comment]
Hi.

I'm have this problem when using docker using `"containerd-snapshotter": true` with `docker compose build`.

errors:
```
target <service>: failed to solve: failed to mount /var/lib/containerd/tmpmounts/containerd-mount2996128479: mount source: "rpool/containerd/362", target: "/var/lib/containerd/tmpmounts/containerd-mount2996128479", fstype: zfs, flags: 1, data: "", err: device or resource busy

level=error msg=/moby.buildkit.v1.Control/Solve error="rpc error: code = Unknown desc = failed to mount /var/lib/containerd/tmpmounts>


...
last of this errors:
...
time="2025-11-23T12:28:56.540817957+01:00" level=error msg="(*service).Write failed" error="rpc error: code = Unavailable desc = ref moby/1/layer-sha25620043066d3d5c78b45520c5707319835ac7d1f3d7f0dded0138ea0897d6a3188 locked for 16.695154862s (since 2025-11-23 12:28:39.845640072 +0100 CET m=+22.305092773): unavailable" expected="sha256:20043066d3d5c78b45520c5707319835ac7d1f3d7f0dded0138ea0897d6a3188" ref="layer-sha256:20043066d3d5c78b45520c5707319835ac7d1f3d7f0dded0138ea0897d6a3188" total=29724688
```
docker info:
```
Client: Docker Engine - Community
 Version:           29.0.2
 API version:       1.52
 Go version:        go1.25.4
 Git commit:        8108357
 Built:             Mon Nov 17 12:33:35 2025
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          29.0.2
  API version:      1.52 (minimum version 1.44)
  Go version:       go1.25.4
  Git commit:       e9ff10b
  Built:            Mon Nov 17 12:33:35 2025
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v2.1.5
  GitCommit:        fcd43222d6b07379a4be9786bda52438f0dd16a1
 runc:
  Version:          1.3.3
  GitCommit:        v1.3.3-0-gd842d771
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```
containerd infos:
`containerd containerd.io v2.1.5 fcd43222d6b07379a4be9786bda52438f0dd16a1`

docker config:
```
{
  "storage-driver": "zfs",
  "features": {
    "containerd-snapshotter": true
  }
}
```

containerd config:
```
[plugins.'io.containerd.cri.v1.images']
  snapshotter = "zfs"
```

os version:
`Debian Trixie`

zfs version:
`zfs-2.3.2-2`

zfs pool:
```
zpool create \
    -o ashift=12 \
    -o autotrim=on \
    -O acltype=posixacl -O xattr=sa -O dnodesize=auto \
    -O compression=lz4 \
    -O normalization=formD \
    -O relatime=on \
    -O canmount=off -O mountpoint=/ -R /mnt \
    rpool mirror\
    /dev/disk/by-id/<name>-part4 \
    /dev/disk/by-id/<name>-part4 \
```

zfs datasets:
```
zfs create -o com.sun:auto-snapshot=false rpool/var/lib/docker
zfs create -o com.sun:auto-snapshot=false rpool/var/lib/containerd
zfs create -o mountpoint=/var/lib/containerd/io.containerd.snapshotter.v1.zfs rpool/containerd
```

I don't have this problem when using:
```
{
  "storage-driver": "zfs",
  "features": {
    "containerd-snapshotter": false
  }
}
```

**Comment by chrisstornowski** [comment]
Since ZFS version 2.2 supports overlay, I've been using it. Everything looks good.
